GWFish/modules/waveforms.py

```diff
@@ -113,51 +130,308 @@ def t_of_f_PN(parameters, frequencyvector):
 
     return t_of_f+parameters['geocent_time']
 
-def lal_caller(waveform, frequencyvector, mass_1, mass_2, luminosity_distance, redshift, theta_jn, phase, geocent_time,
-           a_1=0, tilt_1=0, phi_12=0, a_2=0, tilt_2=0, phi_jl=0, lambda_1=0, lambda_2=0, **kwargs):
-    params_lal = lal.CreateDict()
-    approx_lal = lalsim.GetApproximantFromString(waveform)
-
-    if lambda_1 != 0:
-        from lalsimulation import SimInspiralWaveformParamsInsertTidalLambda1
-        SimInspiralWaveformParamsInsertTidalLambda1(params_lal, float(lambda_1))
-    if lambda_2 != 0:
-        from lalsimulation import SimInspiralWaveformParamsInsertTidalLambda2
-        SimInspiralWaveformParamsInsertTidalLambda2(params_lal, float(lambda_2))
-
-    frequencyvector = frequencyvector.copy().flatten()
-
-    iota, spin_1x, spin_1y, spin_1z, spin_2x, spin_2y, spin_2z = bilby_to_lalsimulation_spins(
-        theta_jn=theta_jn, phi_jl=phi_jl, tilt_1=tilt_1, tilt_2=tilt_2,
-        phi_12=phi_12, a_1=a_1, a_2=a_2, mass_1=mass_1, mass_2=mass_2,
-        reference_frequency=50., phase=phase)
-
-    frequency_array = CreateREAL8Vector(len(frequencyvector))
-    frequency_array.data = frequencyvector
-
-    h_plus, h_cross = lalsim.SimInspiralChooseFDWaveformSequence(
-        phase,
-        mass_1 * lal.MSUN_SI * (1 + redshift),  # in [kg]
-        mass_2 * lal.MSUN_SI * (1 + redshift),  # in [kg]
-        spin_1x, spin_1y, spin_1z, spin_2x, spin_2y, spin_2z,
-        50.,  # reference frequency
-        luminosity_distance * lal.PC_SI * 1e6,  # in [m]
-        iota,
-        params_lal,
-        approx_lal,
-        frequency_array
-    )
-
-    # Add initial 2pi*f*tc - phic - pi/4 to phase
-    phi_in = np.exp(1.j*(2*frequencyvector*np.pi*geocent_time))
-    hp = phi_in * np.conjugate(h_plus.data.data)  # it's already multiplied by the phase
-    hc = phi_in * np.conjugate(h_cross.data.data)
-
-    hp = hp[:, np.newaxis]
-    hc = hc[:, np.newaxis]
-    polarizations = np.hstack((hp, hc))
 
-    return polarizations
+class Waveform:
+    def __init__(self, name, gw_params, data_params):
+        self.name = name
+        self._set_default_gw_params()
+        self.gw_params.update(gw_params)
+        self.data_params = data_params
```

Comment:
Do we really need this layer of abstraction here? if the parameters are just `frequencyvector` and `delta_t` maybe we can just pass them explicitly...

Reply:
I think it would be useful to have more general categories. E.g., for Ulyana's comment I added `f_ref` as another parameter.

Reply:
Also, I was thinking ahead for this removal of `hphc_caller`, and with this in mind, and now it is convenient for all waveform classes to be initiated with the same set of arguments.

Reply:
OK, sounds fair!